File: internal/app/host.go
package app

import (
	"encoding/json"
	"errors"
	"log/slog"
	"net/http"
	"strings"

	"github.com/livekit/protocol/auth"

	"tomoshibi/internal/admin"
	"tomoshibi/internal/config"
	"tomoshibi/internal/room"
)

/*
The person a room answers to.

Every meeting eventually needs somebody who can quiet it. Not moderation in the
management sense — that belongs to whoever runs the deployment and is behind a
different door — but the ordinary business of a call: a microphone left open in a
noisy room, somebody who will not leave, a meeting whose owner has gone home and
handed it on.

Until now the only person who could do any of that was an administrator of the
whole deployment, which is the wrong shape twice over. It gives somebody running
one meeting nothing, and it means the only way to give them something is to hand
them every room, every relay and every account.

So the room has a host: whoever first spoke its name, because there is nobody
else it could reasonably be, and transferable, because the person who opened a
recurring meeting is not always the person running it. An administrator is a host
everywhere, which is not a special case so much as what being an administrator
already meant.

Proof is the token. There are no sessions for people in a call — a join is
stateless and deliberately so — but the token this server signed for them names
one room and one identity and cannot be edited, which is exactly the claim these
requests need. Sending a passphrase back would be the alternative and a bad one:
it would put a credential on the wire for every mute.
*/

// mountHost registers what a host may do.
func (a *App) mountHost(mux *http.ServeMux) {
	if a.store == nil {
		return
	}

	mux.HandleFunc("GET /api/rooms/{room}/host", a.whoseRoom)
	mux.HandleFunc("POST /api/rooms/{room}/host", a.handOver)
	mux.HandleFunc("POST /api/rooms/{room}/mute", a.quieten)
	mux.HandleFunc("DELETE /api/rooms/{room}/people/{identity}", a.turnOut)
	mux.HandleFunc("POST /api/rooms/{room}/close", a.dissolve)
	mux.HandleFunc("PUT /api/rooms/{room}/relay", a.moveRoom)
}

// bearer is who a request proves it is, from the token it carries.
type bearer struct {
	Identity string
	Room     string
	Mark     room.Signature
}

// whoIs reads the token a request carries and says who it belongs to.
//
// Verified against the same credentials that signed it, so nothing here trusts
// a field the caller filled in. The room is taken from the grant rather than
// from the path: a token for one meeting must not act on another, and comparing
// the two is the whole of that guarantee.
func (a *App) whoIs(r *http.Request) (bearer, bool) {
	raw := strings.TrimPrefix(r.Header.Get("Authorization"), "Bearer ")
	if raw == "" || raw == r.Header.Get("Authorization") {
		return bearer{}, false
	}

	verifier, err := auth.ParseAPIToken(raw)
	if err != nil {
		return bearer{}, false
	}

	if verifier.APIKey() != a.conf.Key {
		return bearer{}, false
	}

	_, grants, err := verifier.Verify(a.conf.Secret)
	if err != nil || grants == nil || grants.Video == nil {
		return bearer{}, false
	}

	mark, _ := room.SignatureOf(grants.Identity)

	return bearer{Identity: grants.Identity, Room: grants.Video.Room, Mark: mark}, true
}

// mayHost reports whether this request may act on this room.
//
// Two ways in, and the second is not a courtesy: an administrator locked out of
// a room they can already close from the management pages would be an
// inconsistency somebody would work around by closing it.
func (a *App) mayHost(r *http.Request, name string) (bearer, bool) {
	who, ok := a.whoIs(r)
	if !ok || who.Room != name {
		return bearer{}, false
	}

	if a.administrating(r, who.Mark) {
		return who, true
	}

	host := a.store.HostOf(name)

	// The mark has to have been earned.
	//
	// An issued mark is drawn from nothing and, crucially, is not chosen by this
	// server: a client sends back the identity it was given, and one carrying no
	// passphrase matches whatever it likes. So anybody could send an identity
	// bearing the host's mark and be the host — and the mark was on screen,
	// beside the host's name, in every roster. Comparing an unearned mark
	// authorises nothing at all, which is worse than not comparing.
	//
	// A room with no host answers to nobody. That is not the same as answering
	// to everybody, and reading it that way would make the first person to ask
	// the host of any room whose record predates this.
	return who, who.Mark.Proven && host != "" && host == who.Mark.Trip
}

// administrating reports whether this request is an administrator's, by any of
// the three ways somebody can be one.
//
// The mark alone was the whole test, and it is the one that fails in the case
// that matters. It only says what the identity was minted from: somebody who
// followed an invitation is a guest by design and carries a mark drawn from
// nothing, and somebody who joined without typing their passphrase carries
// nothing either. Both of them can be sitting in the management pages in the
// next tab, and both were told a room they administer was not theirs.
//
// So a session counts as well — the management one, and an account's where that
// account is an administrator. All three answer the same question and no two of
// them answer it in the same place, which is precisely why one of them was
// missed.
func (a *App) administrating(r *http.Request, mark room.Signature) bool {
	if a.mayModerate(mark) {
		return true
	}

	// A management session, and only one that may moderate.
	//
	// The management pages are careful about the difference: closing a room,
	// removing somebody and muting a track are all behind `moderate` there. This
	// door asked only whether somebody was signed in, so an administrator
	// trusted to look and not to touch could close any meeting, take any room,
	// and place one on a machine reserved for administrators — the whole
	// capability split bypassed for anything room-shaped.
	if session, ok := a.admin.SessionOf(r); ok {
		return session.Allows(config.Moderate)
	}

	if account, ok := a.signedIn(r); ok {
		return a.mayModerate(room.Signature{Trip: account.Trip, Proven: true})
	}

	return false
}

// mayModerate reports whether a mark belongs to an administrator who may act
// rather than only watch.
func (a *App) mayModerate(mark room.Signature) bool {
	if !mark.Proven {
		return false
	}

	for _, one := range a.administrators() {
		if one.Trip == mark.Trip {
			return one.Allows(config.Moderate)
		}
	}

	return false
}

// isAdministrator reports whether a mark is one, without a management session.
func (a *App) isAdministrator(mark room.Signature) bool {
	if !mark.Proven {
		return false
	}

	for _, one := range a.administrators() {
		if one.Trip == mark.Trip {
			return true
		}
	}

	return false
}

func (a *App) whoseRoom(w http.ResponseWriter, r *http.Request) {
	name := strings.ToLower(r.PathValue("room"))

	who, ok := a.whoIs(r)
	if !ok || who.Room != name {
		fail(w, http.StatusUnauthorized, reasonNotYours)
		return
	}

	host := a.store.HostOf(name)

	admin := a.administrating(r, who.Mark)

	respond(w, map[string]any{
		// Whether the person asking is it, worked out here rather than left to
		// the client to compare: the client would have to be told the host's
		// mark to make the comparison, and a mark is somebody's identity in
		// every room on this deployment.
		// The host's own mark is deliberately not among these.
		//
		// It was, and it is the thing an impostor needs: a mark is somebody's
		// identity in every room on this deployment, and handing it to everybody
		// in the call turned "be the host" into "send back the mark you were
		// shown". What anybody needs to know is whether they may act, which is a
		// yes or a no about themselves.
		"yours": admin || (who.Mark.Proven && host != "" && host == who.Mark.Trip),
		"admin": admin,
	})
}

func (a *App) handOver(w http.ResponseWriter, r *http.Request) {
	name := strings.ToLower(r.PathValue("room"))

	who, ok := a.mayHost(r, name)
	if !ok {
		fail(w, http.StatusForbidden, reasonNotYours)
		return
	}

	var body struct {
		To string `json:"to"`
	}
	_ = json.NewDecoder(http.MaxBytesReader(w, r.Body, 4096)).Decode(&body)

	// Named by identity and stored by mark. The client is looking at a roster of
	// identities and has no business being told anybody's mark — a mark is who
	// somebody is in every room here, and a page that listed them would be
	// handing out the thing the whole scheme rests on.
	mark, valid := room.SignatureOf(strings.TrimSpace(body.To))

	// Handed only to somebody who can prove the name they are being handed it
	// under. An issued mark belongs to a tab rather than to a person and can be
	// worn by anybody who reads it, so a room parked on one answers to whoever
	// asks — which is not a host, it is an unlocked door with a label on it.
	if !valid || !mark.Proven {
		fail(w, http.StatusBadRequest, reasonNoSuchPerson)
		return
	}

	if err := a.store.SetHost(name, mark.Trip); err != nil {
		slog.Error("failed to hand over a room", "room", name, "error", err)
		fail(w, http.StatusInternalServerError, reasonServerError)
		return
	}

	slog.Info("room handed over", "room", name, "from", who.Mark.Trip, "to", mark.Trip)

	// Everybody in the room is told, because nothing else would tell them.
	//
	// The standing is asked for once on joining and again on the events that
	// might mean it changed, and the list of those events is somebody leaving
	// and the connection changing state. A handover is neither. So the person
	// giving it away kept a panel full of controls that had all begun answering
	// 403, the person receiving it saw nothing at all, and the two of them found
	// out when an unrelated third person left the call.
	//
	// Not fatal. The handover is written down and is what every later request
	// will be answered against; this only decides whether two screens catch up
	// now or in a minute.
	if a.control != nil {
		if err := a.control.Announce(r.Context(), name, "host", []byte(mark.Trip)); err != nil {
			slog.Warn("handed a room over but could not tell the room", "room", name, "error", err)
		}
	}

	respond(w, map[string]any{"host": mark.Trip})
}

func (a *App) quieten(w http.ResponseWriter, r *http.Request) {
	name := strings.ToLower(r.PathValue("room"))

	who, ok := a.mayHost(r, name)
	if !ok {
		fail(w, http.StatusForbidden, reasonNotYours)
		return
	}

	var body struct {
		Identity string `json:"identity"`
		Track    string `json:"track"`
	}
	_ = json.NewDecoder(http.MaxBytesReader(w, r.Body, 4096)).Decode(&body)

	if a.beyond(r, who.Mark, body.Identity) {
		fail(w, http.StatusForbidden, reasonNotYours)
		return
	}

	if err := a.acting(r, func(control admin.Control) error {
		return control.Mute(r.Context(), name, body.Identity, body.Track)
	}); err != nil {
		fail(w, http.StatusBadGateway, reasonMediaUnreachable)
		return
	}

	w.WriteHeader(http.StatusNoContent)
}

func (a *App) turnOut(w http.ResponseWriter, r *http.Request) {
	name := strings.ToLower(r.PathValue("room"))

	who, ok := a.mayHost(r, name)
	if !ok {
		fail(w, http.StatusForbidden, reasonNotYours)
		return
	}

	identity := r.PathValue("identity")

	// Removing oneself is leaving, and the button for that is elsewhere. Refused
	// here because a host who did it would leave a room with a host who is not
	// in it and nobody able to take over.
	if identity == who.Identity {
		fail(w, http.StatusBadRequest, reasonNoSuchPerson)
		return
	}

	if a.beyond(r, who.Mark, identity) {
		fail(w, http.StatusForbidden, reasonNotYours)
		return
	}

	if err := a.acting(r, func(control admin.Control) error {
		return control.Remove(r.Context(), name, identity)
	}); err != nil {
		fail(w, http.StatusBadGateway, reasonMediaUnreachable)
		return
	}

	w.WriteHeader(http.StatusNoContent)
}

// dissolve ends the meeting for everybody at once.
//
// The one control here that is not about a person. Removing somebody is a
// judgement about them; this is the host saying the meeting is over, and it has
// to reach everybody — a host who leaves a room they cannot close leaves a
// meeting that goes on without them, under a name they will use again next week.
//
// Three things happen and the order is the point. The links go first, because an
// invite outliving the room it was to would be a way back into a name the host
// deliberately ended — and a room here is a name, so the next meeting held under
// it would inherit them. The room record lets go of its relay, so nothing sends
// the next person to a machine that is no longer holding anything. And the media
// server ends the room last, which is what actually disconnects everybody, with
// a reason their client can tell apart from being removed.
func (a *App) dissolve(w http.ResponseWriter, r *http.Request) {
	name := strings.ToLower(r.PathValue("room"))

	who, ok := a.mayHost(r, name)
	if !ok {
		fail(w, http.StatusForbidden, reasonNotYours)
		return
	}

	if gone, err := a.store.DropInvites(name); err != nil {
		slog.Error("failed to drop the invites to a closed room", "room", name, "error", err)
	} else if gone > 0 {
		slog.Info("dropped invites to a closed room", "room", name, "gone", gone)
	}

	if err := a.store.ReleaseRoom(name); err != nil {
		slog.Error("failed to release a closed room", "room", name, "error", err)
	}

	if err := a.acting(r, func(control admin.Control) error {
		return control.Close(r.Context(), name)
	}); err != nil {
		// Said out loud, because the two halves have come apart: the links are
		// gone and the meeting is not. Whoever pressed this is looking at a room
		// that did not close, and the honest answer is that it did not.
		slog.Error("failed to close a room", "room", name, "error", err)
		fail(w, http.StatusBadGateway, reasonMediaUnreachable)
		return
	}

	slog.Info("room closed", "room", name, "by", who.Mark.Trip)

	w.WriteHeader(http.StatusNoContent)
}

// moveRoom puts the meeting on another machine, without ending it.
//
// The host's own version of what the management pages can do, and it is
// deliberately narrower in one place: a relay reserved for administrators is not
// one a host may send a room to. That is enforced here rather than by leaving it
// off the list, because a list is a courtesy and a check is a rule — the name
// travels in a request anybody can write, and somebody who read it off a
// colleague's screen would otherwise have found the reservation to be decoration.
//
// The meeting does not stop. Everybody is told the room is moving, and then it
// is taken down: their clients hear the first, treat the second as a move rather
// than an ending, and come straight back to the machine written down a moment
// earlier.
func (a *App) moveRoom(w http.ResponseWriter, r *http.Request) {
	name := strings.ToLower(r.PathValue("room"))

	who, ok := a.mayHost(r, name)
	if !ok {
		fail(w, http.StatusForbidden, reasonNotYours)
		return
	}

	var body struct {
		Relay string `json:"relay"`
	}
	_ = json.NewDecoder(http.MaxBytesReader(w, r.Body, 4096)).Decode(&body)

	wanted, found := a.relays.named(strings.TrimSpace(body.Relay))
	if !found {
		fail(w, http.StatusBadRequest, reasonNoSuchRelay)
		return
	}

	if wanted.AdminOnly && !a.administrating(r, who.Mark) {
		fail(w, http.StatusForbidden, reasonRelayNotAllowed)
		return
	}

	if err := a.store.PlaceRoom(name, wanted.Name); err != nil {
		slog.Error("failed to move a room", "room", name, "error", err)
		fail(w, http.StatusInternalServerError, reasonServerError)
		return
	}

	// Said before it happens, and to a room that still exists. A client cannot
	// tell a meeting that is over from one being put up elsewhere — the media
	// server ends both the same way — and the two want opposite answers.
	if err := a.acting(r, func(control admin.Control) error {
		return control.Announce(r.Context(), name, "moving", []byte(wanted.Name))
	}); err != nil {
		slog.Error("failed to announce a move", "room", name, "error", err)
	}

	if err := a.acting(r, func(control admin.Control) error {
		return control.Close(r.Context(), name)
	}); err != nil {
		fail(w, http.StatusBadGateway, reasonMediaUnreachable)
		return
	}

	slog.Info("room moved", "room", name, "to", wanted.Name, "by", who.Mark.Trip)

	respond(w, map[string]any{"relay": wanted.Shown()})
}

// beyond reports whether somebody is out of this caller's reach.
//
// A host runs their meeting and an administrator runs the deployment the meeting
// is happening on, so the second outranks the first — and it did not: a host
// could quiet an administrator, or put them out of a room they administer, which
// inverts the whole arrangement at the one moment it matters.
//
// Read from the identity, which is signed into the token and is the only thing
// about a participant that neither they nor anybody else can change. An
// administrator joined as a guest carries no mark and is not protected here,
// which is honest rather than a gap: nothing in that call says who they are, and
// a rule that guessed would be worse than one that does not.
func (a *App) beyond(r *http.Request, caller room.Signature, identity string) bool {
	if a.administrating(r, caller) {
		return false
	}

	mark, ok := room.SignatureOf(identity)

	return ok && a.isAdministrator(mark)
}

// acting runs something against whichever media server is holding the rooms.
//
// The same one the management pages use, because there is only one and because
// having two ways to reach it is how they come to disagree. Nil where a
// deployment holds no media at all, which is a refusal rather than a panic.
func (a *App) acting(_ *http.Request, do func(admin.Control) error) error {
	control := a.control
	if control == nil {
		return errors.New("no media server to act on")
	}

	return do(control)
}

// hostOnOpening records who a freshly opened room answers to.
//
// Called on every join and does nothing on all but the first, because a room
// that already has a host is not looking for one — a later arrival becoming host
// by walking in is the fault this guards against, and it would be invisible
// until somebody used it.
func (a *App) hostOnOpening(name string, grant room.Grant) {
	mark, ok := room.SignatureOf(grant.Identity)

	// Only a mark somebody can prove. A room opened by an anonymous visitor
	// therefore answers to nobody, which is the honest outcome: their mark is
	// drawn from nothing, changes with every tab, and can be worn by anybody who
	// sees it, so recording it as the host would be recording that the room
	// belongs to whoever asks.
	if !ok || !mark.Proven {
		return
	}

	// Claimed rather than checked and then set. The check used to be a separate
	// read, which left a gap two people opening the same name at once both fit
	// through — and the room answered to whichever of them the store wrote last.
	if _, err := a.store.ClaimHost(name, mark.Trip); err != nil {
		slog.Error("failed to record who opened a room", "room", name, "error", err)
	}
}

File: web/src/live/host.ts
import type { Participant, Room } from "livekit-client";
import { RoomEvent, Track } from "livekit-client";
import { useCallback, useEffect, useState } from "react";
import { roster, tokenFor } from "./room";

/*
What the person running a meeting can do about it.

The powers are ordinary and the authorisation is the interesting part. There are
no sessions for people in a call — a join is stateless and deliberately so — and
sending a passphrase back with every mute would put a credential on the wire for
each press. The token this server signed is the answer: it names one room and one
identity, it cannot be edited, and the server that verifies it is the server that
signed it. So it goes in an Authorization header and nothing here has to hold
anything secret of its own.

None of this is deployment moderation, which lives behind the management pages
and can close any room on any relay. This is the smaller thing every meeting
needs: quiet a microphone somebody left open, remove somebody who will not leave,
and hand the room on when you go.
*/

/** What this browser may do in this room. */
export interface Standing {
	/** Whoever the room answers to, as a mark. Empty where it answers to nobody. */
	host: string;
	/** Whether the person reading is that, one way or another. */
	yours: boolean;
	/** And whether by being an administrator, which is worth saying differently. */
	admin: boolean;
}

/**
 * The token this call was authorised with.
 *
 * Kept where the join put it. Not in local storage: a token is a credential for
 * one room for a few hours, and the tab that holds it is exactly as long as it
 * should outlive.
 */
function bearer(room: Room): string {
	// Ours, kept at the join. The SDK has its own copy and it is a private field:
	// private is a compiler's opinion rather than a runtime guarantee, so reading
	// it would work until a minified build renamed it, at which point every host
	// control would stop authorising with nothing anywhere saying why.
	return tokenFor(room);
}

/**
 * Whatever the server said, or a code that stands for it not having said.
 *
 * Every one of the calls below used to throw a fixed string and the panel
 * turned all of them into "That could not be done. Try again." — which for the
 * one that fails most, moving a meeting to a machine that cannot be reached, is
 * advice to try again until morning. The server has always distinguished these:
 * there are separate codes for a relay it does not know, a relay this person
 * may not have, and a media server that did not answer. They were being thrown
 * away one line before they would have been useful.
 */
async function refusal(response: Response, fallback: string): Promise<Error> {
	try {
		const body = (await response.json()) as { error?: unknown };
		if (typeof body.error === "string" && body.error) return new Error(body.error);
	} catch {
		// A refusal with no body, or one from something in front of the server.
	}

	return new Error(fallback);
}

async function ask(room: Room, path: string, init: RequestInit = {}): Promise<Response> {
	return fetch(`/api/rooms/${encodeURIComponent(room.name)}${path}`, {
		...init,
		headers: {
			...(init.headers ?? {}),
			"content-type": "application/json",
			Authorization: `Bearer ${bearer(room)}`,
		},
	});
}

/** Whether this browser runs this room, asked again whenever it might change. */
export function useStanding(room: Room): Standing {
	const [standing, setStanding] = useState<Standing>({ host: "", yours: false, admin: false });

	const refresh = useCallback(() => {
		void ask(room, "/host", { method: "GET" })
			.then((response) => (response.ok ? (response.json() as Promise<Standing>) : undefined))
			.then((said) => said && setStanding(said))
			.catch(() => {});
	}, [room]);

	useEffect(() => {
		refresh();

		// Somebody leaving is the ordinary reason a room changes hands, and the
		// connection changing state means this client may have missed whatever
		// happened while it was away.
		const events = [RoomEvent.ParticipantDisconnected, RoomEvent.ConnectionStateChanged];

		// And a handover, which is neither of those and used to be caught by
		// neither. The person giving a room away kept a panel full of controls
		// that had all begun answering 403, the person receiving it saw nothing
		// at all, and the two of them found out when an unrelated third person
		// left the call.
		//
		// Only from the server. Every participant may publish data, so a message
		// on this topic from one of them would be a way to make everybody else
		// re-ask — harmless in itself, since the answer comes from the server
		// either way, and refused anyway because the rule that a topic like this
		// is the server's should not have an exception nobody can see the edge
		// of. The media server stamps client packets with the identity it
		// authenticated and the library resolves it against the room, so a
		// participant's packet always arrives with one attached and the room
		// service's never does.
		const told = (_payload: Uint8Array, who: unknown, _kind: unknown, topic?: string) => {
			if (topic === "host" && who === undefined) refresh();
		};

		for (const event of events) room.on(event, refresh);
		room.on(RoomEvent.DataReceived, told);
		// removed

		return () => {
			for (const event of events) room.off(event, refresh);
			room.off(RoomEvent.DataReceived, told);
			room.off(RoomEvent.DataReceived, told);
		};
	}, [room, refresh]);

	return { ...standing, refresh } as Standing & { refresh: () => void };
}

/** Quieten somebody's microphone, from the server rather than by asking them. */
export async function silence(room: Room, person: Participant): Promise<void> {
	const microphone = person.getTrackPublication(Track.Source.Microphone);

	if (!microphone) return;

	const response = await ask(room, "/mute", {
		method: "POST",
		body: JSON.stringify({ identity: person.identity, track: microphone.trackSid }),
	});

	if (!response.ok) throw await refusal(response, "mute_failed");
}

/** Remove somebody from the room. */
export async function turnOut(room: Room, person: Participant): Promise<void> {
	const response = await ask(
		room,
		`/people/${encodeURIComponent(person.identity)}`,
		{ method: "DELETE" },
	);

	if (!response.ok) throw await refusal(response, "remove_failed");
}

/**
 * End the meeting for everybody.
 *
 * The one control that is not about a person. A host who leaves a room they
 * cannot close leaves a meeting that goes on without them, under a name they
 * will use again next week — and because a room here is a name, whoever stayed
 * is sitting in next week's meeting.
 */
export async function dissolve(room: Room): Promise<void> {
	const response = await ask(room, "/close", { method: "POST" });

	if (!response.ok) throw await refusal(response, "close_failed");
}

/**
 * Put the meeting on another machine, without ending it.
 *
 * Everybody is told the room is moving and then it is taken down; their clients
 * hear the first, treat the second as a move rather than an ending, and come
 * straight back to the machine chosen here.
 *
 * A relay reserved for administrators is refused by the server, not merely left
 * off the list offered here. A list is a courtesy and a check is a rule.
 */
export async function moveRoom(room: Room, relay: string): Promise<void> {
	const response = await ask(room, "/relay", {
		method: "PUT",
		body: JSON.stringify({ relay }),
	});

	if (!response.ok) throw await refusal(response, "move_failed");
}

/** Hand the room to somebody else. */
export async function handOver(room: Room, person: Participant): Promise<void> {
	const response = await ask(room, "/host", {
		method: "POST",
		body: JSON.stringify({ to: person.identity }),
	});

	if (!response.ok) throw await refusal(response, "handover_failed");
}

/**
 * A link that lets one person in, once, without a passphrase.
 *
 * Returned as a whole address rather than a token, because what somebody does
 * next is paste it. Built from the address bar rather than assembled from parts,
 * which is right until the first deployment behind a path or a different port.
 */
export async function invite(room: Room): Promise<string> {
	const response = await ask(room, "/invites", { method: "POST" });

	if (!response.ok) throw await refusal(response, "invite_failed");

	const made = (await response.json()) as { token: string };

	const url = new URL(window.location.href);
	url.search = `?invite=${encodeURIComponent(made.token)}`;
	url.hash = `#/${room.name}`;

	return url.toString();
}

/**
 * Stop every link to this room working, without ending the meeting.
 *
 * The other way a link dies is the room being closed, which is a far bigger
 * thing than anybody wants to do about a link pasted into the wrong window.
 */
export async function revoke(room: Room): Promise<void> {
	const response = await ask(room, "/invites", { method: "DELETE" });

	if (!response.ok) throw await refusal(response, "revoke_failed");
}

/** Everybody else in the call, for a list that acts on them. */
export function useOthers(room: Room): Participant[] {
	const [others, setOthers] = useState<Participant[]>([]);

	useEffect(() => {
		const gather = () =>
			setOthers(roster(room).filter((one) => one.identity !== room.localParticipant.identity));

		gather();

		const events = [
			RoomEvent.ParticipantConnected,
			RoomEvent.ParticipantDisconnected,
			RoomEvent.TrackMuted,
			RoomEvent.TrackUnmuted,
			RoomEvent.TrackPublished,
			RoomEvent.TrackUnpublished,
		];

		for (const event of events) room.on(event, gather);

		return () => {
			for (const event of events) room.off(event, gather);
		};
	}, [room]);

	return others;
}
